internal/generate/handler.go

```diff
@@ -34,166 +36,160 @@ import (
 )
 
 func NewHandler(ctx context.Context, config *Config, env *serverenv.ServerEnv) (http.Handler, error) {
-	logger := logging.FromContext(ctx)
+	logger := logging.FromContext(ctx).Named("generate")
 
 	if env.Database() == nil {
 		return nil, fmt.Errorf("missing database in server environment")
 	}
 
-	transformer, err := model.NewTransformer(config)
+	transformer, err := publishmodel.NewTransformer(config)
 	if err != nil {
 		return nil, fmt.Errorf("model.NewTransformer: %w", err)
 	}
-	logger.Infof("max keys per upload: %v", config.MaxKeysOnPublish)
-	logger.Infof("max interval start age: %v", config.MaxIntervalAge)
-	logger.Infof("truncate window: %v", config.TruncateWindow)
+	logger.Debugw("max keys per upload", "value", config.MaxKeysOnPublish)
+	logger.Debugw("max interval start age", "value", config.MaxIntervalAge)
+	logger.Debugw("truncate window", "value", config.TruncateWindow)
 
 	return &generateHandler{
 		serverenv:   env,
 		transformer: transformer,
 		config:      config,
-		database:    database.New(env.Database()),
+		database:    publishdb.New(env.Database()),
+		logger:      logger,
 	}, nil
 }
 
 type generateHandler struct {
 	config      *Config
 	serverenv   *serverenv.ServerEnv
-	transformer *model.Transformer
-	database    *database.PublishDB
+	transformer *publishmodel.Transformer
+	database    *publishdb.PublishDB
+	logger      *zap.SugaredLogger
 }
 
 func (h *generateHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	ctx, span := trace.StartSpan(r.Context(), "(*generate.generateHandler).ServeHTTP")
 	defer span.End()
-	logger := logging.FromContext(ctx)
+
+	logger := h.logger.Named("ServeHTTP")
 
 	regionStr := h.config.DefaultRegion
-	if regionParams, ok := r.URL.Query()["region"]; ok && len(regionParams) > 0 {
-		regionStr = regionParams[0]
+	if v := r.URL.Query().Get("region"); v != "" {
+		regionStr = v
 	}
 	regions := strings.Split(regionStr, ",")
-	logger.Infof("Request to generate data for regions: %v", regions)
+
+	logger.Debugw("generating data", "regions", regions)
+
+	if err := h.generate(ctx, regions); err != nil {
+		logger.Errorw("failed to generate", "error", err)
+		span.SetStatus(trace.Status{Code: trace.StatusCodeInternal, Message: err.Error()})
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(w, err.Error())
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	fmt.Fprint(w, "successfully generated exposure keys")
+}
+
+func (h *generateHandler) generate(ctx context.Context, regions []string) error {
+	logger := h.logger.Named("generate")
+
+	if h.config.KeysPerExposure < 2 {
```

Comment:
why 2?

Reply:
Line 141:

```go
intervalIdx, err := util.RandomInt(len(publish.Keys) - 1)
```

That gets a random index in the list, but panics if the number of keys is 1 (and therefore tries to generate a random int between 0 and 0).

Reply:
worth a comment in the code then